### ISSUE TITLE
Improve the usage of Sysfs in the interaction with EV3Dev

### ISSUE THREAD

**Issue by jabrena** [ev3dev-lang-java/ev3dev-lang-java#765]
In the previous Issue:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/issues/748

We improved the performance with current architecture, the usage of `Sysfs`. `Sysfs` in the current form, is a stateless class with static methods. 

Ideas to improve in this iteration:

- [ ] Moving from static methods, to instance methods
- [ ] Add setSysfs to `Mock` Sysfs to improve testing

@dwalend, what other ideas, do you have?
@JakubVanek, what other ideas, do you have? 

In your opinion, what state is necessary to maintain in `Sysfs` and have some kind of concurrency protection?

**Comment by dwalend** [comment]
I don't think we should try for concurrency safety or state inside Sysfs. I don't think we can do it correctly and do it well at the same time; I think we'd regret any choice internal to Sysfs. 

I think the best path forward is to step away from Sysfs, and start using DataChannelRereaders and (new) DataChannelRewriters directly in Battery, GenericMode and BaseRegulatedMotor (contains EV3LargeRegulatedMotor's interesting code). (That's what I did for my kid's basic control loop.) 

The nice thing about Battery is that it only needs 2 readers, zero writers. BaseRegulatedMotor needed 4 readers and 5 writers. The GyroSensor had 1 writer and 3 readers.

I should have some free evenings this week after the kids are in bed.

I could do a PR of DataChannelRereader in Battery - and make it Closable - as a first step if that's of interest. 

As a second step I can do a PR for a DataChannelRewriter.

**Comment by jabrena** [comment]
Let’s try your idea. 
With Battery, we will explore the model.

Apart of the performance, the first goal in this issue, I am pretty interesting to introduce Mockito. In the project we added tests (One innovation in LEJOS ecosystem), but it is possible to improve.

**Comment by dwalend** [comment]
For 
```
    public float getBatteryCurrent() {
        if (CURRENT_PLATFORM.equals(EV3DevPlatform.EV3BRICK)) {
            return Sysfs.readFloat(BATTERY_PATH + "/" + BATTERY_EV3 + "/" + CURRENT);
        } else {
            LOGGER.warn("This method is not available for {} & {}", EV3DevPlatform.PISTORMS, EV3DevPlatform.BRICKPI);
            return -1f;
        }
    }
```

Does that file path exist on every platform (so the code can open the file) or is it missing from PISTORMS and BRICKPI ?

**Comment by dwalend** [comment]
Also, most of the codebase looks like it is after the "`IOException`s are annoying" revelation. Should I follow that pattern or throw IOExcpetion s?

**Comment by dwalend** [comment]
For Mockito - slipping an alternate filesystem in place is pretty easy. Does Mockito add value to that approach? (I think it at least adds a little.)

**Comment by jabrena** [comment]
Hi @dwalend,

> Does that file path exist on every platform (so the code can open the file) or is it missing from PISTORMS and BRICKPI ?

Exactly, this particular method is only available in EV3 Brick.

> Also, most of the codebase looks like it is after the "IOExceptions are annoying" revelation. Should I follow that pattern or throw IOExcpetion s?

Interesting point for a Debate. Currently I return a bad value and I advice in logs?
Maybe it is more strict to stop the program with a runtime exception and explain why.
What do you think?

> For Mockito - slipping an alternate filesystem in place is pretty easy. Does Mockito add value to that approach? (I think it at least adds a little.)

Mockito introduce an standard in the way to test software as a Best practice. In general, people forget `Testing phase` due to multiple factors and this is a good opportunity to to show the advantage of standard testing in Java but in robotics side.

**Comment by dwalend** [comment]
`Battery` doesn't seem to get any value from inheriting from `EV3DevDevice` or `Power` . Should I pull those out now, or save it for a different change? (My preference is a different change to keep concerns separate, but it does mean something out there could be calling the various set*s and get*s on battery via promoting to public and reflecting.

PR to follow in a moment.

**Comment by dwalend** [comment]
Here's the PR: https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/770

**Comment by dwalend** [comment]
Hmm. Instead of using -1f as the bad value try out Float.NaN . I think in Java Exceptions are a better answer, but at least NaN will blow up if someone tries to multiply or add it. 

I'm coming back to Java from Scala. I don't miss the caught Exceptions like I thought I might. Java's the only language that has them I think. They put you into a contravariant world. But really if you're using the wrong robot then the Battery class isn't going to do what you'd hoped and there's no saving that part of the program.

**Comment by dwalend** [comment]
Here's a draft DataChannelRewriter : https://github.com/dwalend/ev3dev-lang-java/pull/1/files

**Comment by jabrena** [comment]
I have merged the changes. The new version maintains a good readability. I see your strategy, with this kind of inmutable objects, we will increase the performance.

> I think in Java Runtime Exceptions are a better answer, but at least NaN will blow up if someone tries to multiply or add it.

Yes, please, change the code to return a Runtime Exception.

> Battery doesn't seem to get any value from inheriting from EV3DevDevice or Power . 

For EV3DevDevice, it provides some objects like ev3DevProperties
https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/master/src/main/java/ev3dev/hardware/EV3DevDevice.java

Power provides the contract for Battery.
https://github.com/ev3dev-lang-java/lejos-commons/blob/master/src/main/java/lejos/hardware/Power.java

To test the new class: DataChannelRewriter, I believe that the class EV3Led could be a good candidate:

- ev3dev.actuators.ev3.EV3Led

We should have a Unit Test for the class: DataChannelRereader.java

Can you try to Mock: DataChannelRereader from BatteryTest?

**Comment by JakubVanek** [comment]
> inmutable objects

Unfortunately, DataChannelRereader is currently not immutable. The problem is with the cached [`byteBuffer`](https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/c69ceaba8e4860d52b42aaa7f495c5220c46c428/src/main/java/ev3dev/utils/DataChannelRereader.java#L20) member in that class. This also currently breaks multithreading safety - if two threads read from the same Battery instance (which they will, because it is a singleton that cannot be recreated), there is a possibility that two calls to `<rereader>.readString()` will get interleaved in a way that one or both receive incorrect data.

There is also a caveat with the cached FileChannel that I didn't realize earlier: https://stackoverflow.com/a/53361005/1561345 (state/current file pointer is visible across threads and could cause problems too). I am hoping this could be fixed by using a `read()` function with the offset directly specified instead of relying on `position()`.

The two above fields mean that methods using DataChannelRereader are currently safely callable only from one thread at a time.

My intuition is getting OK with that approach, but I'd add a warning to Javadoc for that method about that. For example, `strerror()` in POSIX/C is documented to be racy: https://man7.org/linux/man-pages/man3/strerror.3.html.

**Comment by jabrena** [comment]
Hi @JakubVanek, 

the class is Immutable:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/sysfs_perf2/src/main/java/ev3dev/utils/DataChannelRereader.java

> In object-oriented and functional programming, an immutable object (unchangeable object) is an object whose state cannot be modified after it is created.

https://en.wikipedia.org/wiki/Immutable_object

other stuff is the `Thread Safety` but lets give time to @dwalend to review your comment. 

I put some links as a reference:

- https://stackoverflow.com/questions/11153583/options-to-make-javas-bytebuffer-thread-safe
- https://stackoverflow.com/questions/36974760/how-to-use-bytebuffer-properly-in-multiple-thread-environment

**Comment by JakubVanek** [comment]
I think the non-thread-safe nature is intentional (thread safety decisions are shifted to the user). I just consider this unfortunate with singletons (only one thread should to access Battery at a time, or an additional locking wrapper has to be used by users). It may also change semantics of programs that were built with the assumption that multithreaded access is OK (this might be only an imagined problem though, I have no statistics about how the API is really used).

**Comment by jabrena** [comment]
> I think the non-thread-safe nature is intentional (thread safety decisions are shifted to the user).

No, **ByteBuffer** has a known bug and It was solved in `Java 13`.
https://www.oracle.com/java/technologies/javase/13-relnote-issues.html
https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-5029431

**Comment by JakubVanek** [comment]
Hmmm, regarding the `Power` interface - this is how it worked on real leJOS:
* `Battery` was likely only a compatibility layer, as it is a static class: https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/Battery.java
* It deferred processing to `Power` implementations: https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/hardware/Power.java
* For EV3, this was implemented by an internal `EV3Battery` class: https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/internal/ev3/EV3Battery.java
* The voltage readout itself was implemented in `EV3AnalogPort` as a static function that reads the value a shared memory buffer provided by the kernel driver: https://sourceforge.net/p/lejos/ev3/code/ci/master/tree/ev3classes/src/lejos/internal/ev3/EV3AnalogPort.java#l343

**Comment by JakubVanek** [comment]
> > I think the non-thread-safe nature is intentional (thread safety decisions are shifted to the user).
> 
> No, **ByteBuffer** has a known bug and It was solved in `Java 13`.
> https://www.oracle.com/java/technologies/javase/13-relnote-issues.html
> https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-5029431

This could work, but as the issue notes, the intent is likely the following:

> Consider a use case where multiple threads are to operate on a large region of memory by dividing it up into segments, and having each thread operate on its own segment.

**Comment by jabrena** [comment]
Sorry, @JakubVanek 

Please Stop doing it. 
This issue is not a chat.

**Comment by JakubVanek** [comment]
I am sorry for posting another comment (this time it is related directly to the threading issue). One way out would be to declare the whole library as safe to access only from one thread (IIUC Swing does this similarly, it has its own UI thread and changes can be safely made only from within it). Under that assumption the changes in Battery & DataChannelRereader should work fine and indeed, improve performance. However, I'd consider this an API contract change and as such it should be documented and mentioned in the release notes.

**Comment by jabrena** [comment]
Listen me @JakubVanek,

Don´t say sorry, your ideas are valuable in many many cases, but you need to measure your energy. 
If @dwalend is focusing on this issue, he needs to lead it.

Review the backlog, exist new good issues to develop from scratch new stuff.
https://github.com/ev3dev-lang-java/ev3dev-lang-java/issues

Indeed, Power interface will have more relevance in the future.
https://github.com/ev3dev-lang-java/ev3dev-lang-java/issues/771

Enjoy

Good night

**Comment by dwalend** [comment]
Do you two think it's worth making the read() and close() methods in DataChannelRereader synchronized ? (Anything Closable has at least two states.) That'd give the thread safety. The JVM is pretty good at optimizing synchronized away if only one thread uses a lock. 

I personally think it is reasonable to just mention that the calling code is responsible for concurrent access as a caveat in the docs that this code works like most of the rest of the Java ecosystem. Ultimately there's no reasonable way to prevent two separate processes from writing the same file. We could only ever fix so much of the problem adding locks to a library.

We could always measure performance and let that guide us.

Crazy day today. I'll slip these changes in - likely tomorrow evening.

**Comment by jabrena** [comment]
Hi @dwalend,

I have uploaded a TestCase to review the concurrency for the new class: `DataChannelRereader.java` working in isolation and used by `Sysfs.java`.

https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/sysfs_perf2/src/test/java/ev3dev/utils/DataChannelRereaderConcurrencyTest.java

The testcases are:

```java
    /**
     * Writer1 -> pairs.txt
     * Reader1 <- pairs.txt
     * Reader2 <- pairs.txt
     *
     * Writer1 -> odds.txt
     * Reader1 <- odds.txt
     * Reader2 <- odds.txt
     */
```

This is case, is not very realistic because `ev3dev` will upgrade the file and later `Sysfs` will read the value. At the end the path is unique so the user should share in the program the same object but in this case, I am checking what happen if exist multiple readers.

When I execute multiple times, I receive the following exception in some case:

```
Caused by: java.io.IOException: Premature end of file 
	at ev3dev.utils.DataChannelRereader.readString(DataChannelRereader.java:59)
	... 14 common frames omitted
```

From the block:

``` java
            do {
                byteBuffer.clear();
                channel.position(0);
                n = channel.read(byteBuffer);
                if (n == -1) {
                    throw new IOException("Premature end of file ");
                }
            } while (n <= 0);
```

In the second scenario, it is more realistic:

```java
    /**
     * Writer1 -> pairs.txt
     * Reader1 <- pairs.txt
     *
     * Writer1 -> odds.txt
     * Reader1 <- odds.txt
     */
```

`ev3dev` will upgrade the file and later `Sysfs` will read the value. At the end the path is unique and only exist a reader but again, I receive the the exception: `"Premature end of file "`

Why appear that exception?

Juan Antonio

**Comment by JakubVanek** [comment]
@jabrena I have quickly looked at `DataChannelRereaderConcurrencyTest` and it seems to me that the data race is accidentally avoided there. The reason is that each CompletableFuture has its own DataChannelRereader and thus the buffer is not shared. We cannot achieve this effective per-thread separation with Battery, as threads are not able to create their own Battery instances - they can only get the global singleton instance.

@dwalend Locking doesn't seem strictly necessary. If the buffer is moved to a thread local variable and the file channel is accessed atomically (i.e. only one function call with absolute positioning), the race can be avoided (I think, I may be wrong). I have provided https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/775 with an overview how that could look.

I haven't thought about concurrent writers, the issue I was thinking about is with concurrent readers. There is a vulnerable time window between the moment the NIO read is completed (and data are stored in the buffer) and the moment the data from the buffer are converted to a string. There is also another window between the moment the buffer limits are cleared and the rest of the code.

The scenario that is "haunting" me is that someone tries to read battery voltage from multiple threads (whatever may their motivation be) and they occasionally get an incorrect value and/or exception due to unparseable integer (empty string).

**Comment by jabrena** [comment]
Hi @JakubVanek,

I tested in local your alternative, but I receive the same exception in both tests:

``` java
            int n = channel.read(buffer, 0);
            if (n == -1) {
                throw new IOException("Premature end of file " + path);
            }
```

```
Caused by: java.io.IOException: Premature end of file ./pairs.txt
	at ev3dev.utils.DataChannelRereader2.readString(DataChannelRereader2.java:96)
	... 14 common frames omitted
```

In the second test:

/**
     * Writer1 -> pairs.txt
     * Reader1 <- pairs.txt
     *
     * Writer1 -> odds.txt
     * Reader1 <- odds.txt
     */

exist one reader per file, not concurrent readers. Execute in local the test to review the results.

**Comment by JakubVanek** [comment]
I tried running `./gradlew test` earlier and I cannot get this to fail (OpenJDK 11.0.9.1 from Ubuntu 18.04.5 x86_64). `./gradlew dockerTest` succeeds too. However, I'm wondering if the problem couldn't be having the file open for both reading and writing at the same time (I think Windows has issues with this). I have added https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/775/commits/346425f055b9a26ffb1a45fcffd546aeee10b441, that might fix the problem.

**Comment by jabrena** [comment]
I go for lunch, later I will continue

**Comment by jabrena** [comment]
@JakubVanek, Upgrade the Class: `DataChannelRereader2` with your proposals.

I have improved the tests, you should see that it fails in the CI. (I added some missing asserts)
https://github.com/ev3dev-lang-java/ev3dev-lang-java/actions/runs/452838235

With both Test classes, we can follow the cases.

I have just added a test to review the old Sysfs implementation and I noticed that I receive something similar:

```
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:458)
	at ev3dev.utils.SysfsOld.readString(SysfsOld.java:74)
	at ev3dev.utils.SysfsOld.readInteger(SysfsOld.java:94)
```

``` java
            if (existFile(path) && Files.isReadable(path)) {
                final String result = Files.readAllLines(path, Charset.forName("UTF-8")).get(0);
                if (log.isTraceEnabled()) {
                    log.trace("value: {}", result);
                }
                return result;
            }
```

> The scenario that is "haunting" me is that someone tries to read battery voltage from multiple threads (whatever may their motivation be) and they occasionally get an incorrect value and/or exception due to unparseable integer (empty string).

We need to know the reason that testing only with one Writer and one Reader, the behavior is the same.
For some reason exist something when you write and read (no Concurrent scenario) it fails.

Maybe it is something related to the way that open the file.

**Comment by JakubVanek** [comment]
> * When Java write and Read concurrently, it is possible to have some issues. (Something to tackle, but this is not blocking because users doesn´t have the same object to do different actions, but should find a solution)

I found the issue - `-1` returned from FileChannel.read() actually means that the file is empty (it was - `cat <file>` returned nothing). Therefore the write in Sysfs class is not atomic and there is a moment when the file is empty. This seems consistent with how I'd imagine the syscall sequence:

1. The file is internally opened for writing by Java and truncated (something like `open(path, O_WRONLY | O_CREAT | O_TRUNC, 0664)` in C), thus setting contents to "". The window after this is where DataChannelRereader crashed.
2. Only then the integer/whatever is written to it, making the file contents valid again.

This is going to crash old Sysfs as well, because even if the read succeeds, parsing empty string as an integer will throw another exception.

This also shouldn't affect real sysfs because there is no file content to be truncated. The buffer that is returned in response to read is always freshly generated.

> * When Java Read concurrently, 0 issues with all implementations. (So, we can live with current DataChannelRereader.java)

This may not be the case. The problem is when one shared DataChannelRereader instance is used concurrently. In the concurrency test, multiple separate DataChannelRereaders are used (this works OK). The untested situation where one DataChannelRereader instance is used by two threads happens when these two threads use Battery (which is a singleton with a private constuctor) in parallel and this is where the race is.

> Do we have to raise and exception in this case?

This seems not to be the case, `-1` does not imply error here (in C/POSIX it would). Quoting from FileChannel.read docs ([link](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#read(java.nio.ByteBuffer,%20long))):
> Returns: The number of bytes read, possibly zero, or -1 if the given position is greater than or equal to the file's current size

**Comment by jabrena** [comment]
Good research, 

but how to fix the scenario for concurrent Write/Read?
What is your proposal?

In my opinion, the user doesn´t implement in the same program that scenario. 

I share a concurrent read old example:
https://www.youtube.com/watch?v=3hXTeK8lFMc

**Comment by dwalend** [comment]
I guess my big opening question before examining the test code is: how do you want this to behave concurrently? I think there are two reasonable approaches: synchronized read() and close() or let the caller do the work. I'm fine with either.

The follow-up is: how would you like to handle times when the JVM catches the OS midway through writing a file while the JVM tries to read it? Returning an empty string from readString() seems reasonable in a very local sense, but will still blow up.

Maybe placing readFloat(), readInt(), maybe even readEnum() in DataChannelRereader would give us an easier way to handle what our code does for those "The JVM can't do this now, but try later" scenarios. I'll see what that looks like next.

**Comment by dwalend** [comment]
From https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/channels/ReadableByteChannel.html#read(java.nio.ByteBuffer)

Returns:
The number of bytes read, possibly zero, or -1 if the channel has reached end-of-stream

Shouldn't it return zero for an empty file, and only -1 if the file doesn't exist?

**Comment by jabrena** [comment]
> I guess my big opening question before examining the test code is: how do you want this to behave concurrently? I think there are two reasonable approaches: synchronized read() and close() or let the caller do the work. I'm fine with either. I'm not sure of the value of writing tests of concurrent behavior until we define that expected behavior.

You commented that current approach is followed by the Java Ecosystem

> I personally think it is reasonable to just mention that the calling code is responsible for concurrent access as a caveat in the docs that this code works like most of the rest of the Java ecosystem. 

So, we can continue in current way, so _lets the caller do the work_.

> I'm not sure of the value of writing tests of concurrent behavior until we define that expected behavior.

Review if we have issues in the different scenarios and if some issues detected are scenarios where users use the objects.

**Comment by jabrena** [comment]
I have opened the discussion tab:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/discussions/777

**Comment by dwalend** [comment]
That whole do/while looks suspicious to me... I'm cleaning it out.

**Comment by dwalend** [comment]
Pull request without the do/while at https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/778

**Comment by jabrena** [comment]
Many thanks @dwalend,

Tomorrow, I will review in detail before new year's eve dinner

**Comment by jabrena** [comment]
Many thanks @JakubVanek,

with the alternative based on `ThreadLocal`
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ThreadLocal.html
https://stackoverflow.com/questions/817856/when-and-how-should-i-use-a-threadlocal-variable

**Comment by JakubVanek** [comment]
I have some optimizations for DataChannelRewriter: https://github.com/ev3dev-lang-java/ev3dev-lang-java/commit/7bd73bcfb31a824ed6f4f6c253bd01881b593bb8

* At this stage, the shared buffer is not necessary. The allocations are already made inside `String.getBytes()`, the best thing to do may be to just wrap the returned array in a ByteBuffer. That being said, if custom integer serialization (bypassing strings) is added later, the buffer might be useful again.
* `position()` + `write()` can be folded together - this should be faster & thread-safer (although with `synchronized`, that doesn't matter much).
* `truncate()` can be avoided if the target is in real `/sys`, as it should have no effect there. The truncation is also moved to the beginning of the operation, mimicking the current behaviour of the Sysfs class. This needs some testing though.

As there are two independent file channel operations for each write, I have retained the `synchronized` keyword there. Otherwise, two threads may get interleaved in an unfortunate way and that would produce invalid file contents in emulation (outside of real /sys):

| n | thread A | thread B | file contents |
|-|-|-|-|
| 1 | truncate to zero size | | empty |
| 2 | | truncate to zero size | empty |
| 3 | | write "abcd" to pos 0 | `abcd` |
| 4 | write "xyz" to pos 0 | | `xyzd` |

The result would be "xyzd", which is neither "abcd" or "xyz".

(For comparison, this is what I think would happen in /sys if the truncation was there):
| n | thread A | thread B | kernel action |
|-|-|-|-|
| 1 | truncate to zero size | | noop |
| 2 | | truncate to zero size | noop |
| 3 | | write "abcd" to pos 0 | call attribute store function with "abcd" |
| 4 | write "xyz" to pos 0 | | call attribute store function with "xyz" |


On the other hand, I wouldn't be surprised if "xyzd" occured even with the current Sysfs class, so protecting against this scenario might not be worth it. After all, these are concurrent writes and they already behave strangely. Also, there indeed might be another process changing the file in parallel and there is no easy way of protecting against that (perhaps file locks, but it would have to be mandatory, not just advisory locking).

**Comment by jabrena** [comment]
Returning in 1h from lunch

**Comment by dwalend** [comment]
@JakubVanek - does "outside of real /sys" mean "outside of a real filesystem"  ? Or is there something else special on the ev3dev OS ?

If a file system will do - what do you two think of testing on a file system? It could be a first step in a long journey to an ev3dev simulator.

**Comment by jabrena** [comment]
Hi @dwalend 

> It could be a first step in a long journey to an ev3dev simulator.

The idea is nice, but it is not realistic. Exist many many Mindstorms projects in the past, some of them with a commercial nature and all of them are Ko and it requiere huge mountains of hours. I try to advice you.

**Comment by jabrena** [comment]
I am going to review all PR.

@JakubVanek, I don´t see this commit as a PR:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/commit/7bd73bcfb31a824ed6f4f6c253bd01881b593bb8

**Comment by dwalend** [comment]
@jabrena - I took 7bd73bc as review commentary for my work.  Not sure if that was Jakub's intent.

**Comment by jabrena** [comment]
@JakubVanek, we need to improve the way that you discuss the ideas.

- https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/775/files this PR touch the class from @dwalend, BLOCKED to Merge to compare in local
- https://github.com/ev3dev-lang-java/ev3dev-lang-java/commit/7bd73bcfb31a824ed6f4f6c253bd01881b593bb8 should be promoted as a PR in another class, not the Original class.

Current Status:

- DataChannelRereader supports Multiple Readers, Ok
- DataChannelRereader2 supports Multiple Readers, Ok
- SysfsOld supporst too, but it is so much slower in compare with previous one.

@dwalend In relation to the tests, review the rest of tests, you can "Recreate" a `ev3dev` fs to test the classes. It is not necessary to path DataChannelRereader with if blocks.

**Comment by dwalend** [comment]
@jabrena - what's the right way to recreate the ev3dev file system? 

Battery gets its paths from EV3DevPropertyLoader.getEV3DevProperties() , which is pulling them in from a baked-in property file. I'd like to slip in my own properties file for testing there, but I don't see other code doing that, and don't see a way to do it without altering the EV3DevPropertyLoader API.

There's some kind of inheritance interaction between FakeBattery and BaseElement, but I don't see how that changes the paths that Battery will use.

Is there an example I should follow?

**Comment by JakubVanek** [comment]
> @JakubVanek, we need to improve the way that you discuss the ideas.
> 
> * https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/775/files this PR touch the class from @dwalend, BLOCKED to Merge to compare in local

This PR should be removed, it uses `ThreadLocal<>` which was rejected. However, it also contains `DataChannelRereaderRaceTest` which shows the problematic behavior solved by adding `synchronized` to DataChannelRereader. I can rebase & change the PR to only include that.

> * [7bd73bc](https://github.com/ev3dev-lang-java/ev3dev-lang-java/commit/7bd73bcfb31a824ed6f4f6c253bd01881b593bb8) should be promoted as a PR in another class, not reusing DataChannelRereader

I can create a PR from it, in its current state it was meant as a sample implementation of the ideas in that comment.

**Comment by jabrena** [comment]
@dwalend  SysfsTest.java should be fine

If you feel blocked, upload in a PR the tests and I will help tomorrow. :)

Once we have DataChannelRereader stable and DataChannelRewriter too, we will go faster.

**Comment by jabrena** [comment]
@JakubVanek that version was not rejected. I wanted to have more options.

> This PR should be removed, it uses ThreadLocal<> which was rejected.

but it is true that ThreadLocal<> protect Bytebuffer in a different way but in order to merge, you need to reset the class DataChannelRereader

I would like to compare and observe with `Visualvm` the impact with ThreadLocal

Now, we notice that new versions improve the performance.

**Comment by jabrena** [comment]
For today, is Ok, have a good dinner with the family in USA & Cech Republic

I have to say that this journey is pretty nice and it is not easy.
You should feel happy

**Comment by jabrena** [comment]
Testing in the Brick both alternatives with the concurrent scenario for Readers.

- [ ] DataChannelRereader
- [ ] DataChannelRereader2
- [ ] Sysfs (Old)

Criterias:

- [ ] Performance
- [ ] Memory

Tomorrow, I will share the results.

**Comment by jabrena** [comment]
If you like, tomorrow I will clean the branch to continue with the part about the `Writer`

**Comment by JakubVanek** [comment]
> It is clear, that the usage of DataChannelRereader improve the performance in near of 300% in compare with the usage of Sysfs.

The error ranges are once again huge. For example, peeking at the individual data values shows that the first sample in Throughput is many times larger than the other values. JMH is capable of measuring more accurate results; wouldn't it be possible to adjust the benchmark parameters?

**Comment by dwalend** [comment]
I've got a first pull request at https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/781  . What do you two think of it?

**Comment by jabrena** [comment]
Hi @dwalend, I add waiting for the Unit Test for DataChannelRereader to continue.

**Comment by jabrena** [comment]
Hi @JakubVanek, 

In the performance comparison, you notice that the usage of `DataChannelRereader` offers a better performance and reviewing the solution, the alternative using `ThreadLocal<>` solves the same issue with Concurrent readers but in terms of memory, every object using a threadlocal, forces to have a specific memory area to maintain the Thread Safety so this is the reason that the first solution is better. Better performance, less memory usage and better code understanding. In general, `ThreadLocal` is not the first option but I felt curious to see the performance results but I have to say that the discussion improved the solution for `DataChannelRereader`.

**Comment by JakubVanek** [comment]
> In the performance comparison, you notice that the usage of `DataChannelRereader` offers a better performance and reviewing the solution, the alternative using `ThreadLocal<>` solves the same issue with Concurrent readers but in terms of memory, every object using a threadlocal, forces to have a specific memory area to maintain the Thread Safety so this is the reason that the first solution is better. Better performance, less memory usage and better code understanding. In general, `ThreadLocal` is not the first option but I felt curious to see the performance results but I have to say that the discussion improved the solution for `DataChannelRereader`.

I agree that the synchronized DataChannelRereader solution is better. However, I still think that performance conclusions shouldn't be drawn from the graph posted above due to the error margins that the data has. For example the first outlier sample result shouldn't be counted to the averages; it should be processed separately.

**Comment by jabrena** [comment]
Hi @JakubVanek,

it is interesting. If you are interested in that field, please create a new issue about "Best Practices Performance Testing" and use that issue to add new references and examples using tools from Java Space. It could be pretty valuable to be reused in the future.

**Comment by JakubVanek** [comment]
> it is interesting. If you are interested in that field, please create a new issue about "Best Practices Performance Testing" and use that issue to add new references and examples using tools from Java Space. It could be pretty valuable to be reused in the future.

I feel this is just following common sense, I outlined some of the parts in https://github.com/ev3dev-lang-java/ev3dev-lang-java/issues/748#issuecomment-751272222 and an example of better error ranges (with JMH) is https://github.com/ev3dev-lang-java/ev3dev-lang-java/issues/748#issuecomment-743974830. I have since deleted the forked repository and recreated it, but I can reproduce the configuration if necessary.

**Comment by jabrena** [comment]
Hi @JakubVanek,

It is not necessary.

**Comment by dwalend** [comment]
> Hi @dwalend, I am waiting for the Unit Test for DataChannelRereader to continue.

Will https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/781/files#diff-c0e0309e22ce34fb394a33b423bdafcf7785541728b2b6b6ef21490144a47b38 do? I'm afraid I might have lost the thread.

**Comment by jabrena** [comment]
I will review later the link. 
I am with the family, today It is wise man day

Basically, It is pending a Unit Test class for DataChannelRereader

**Comment by jabrena** [comment]
Hi @dwalend,

I saw the test. I will review tomorrow.

Cheers

**Comment by jabrena** [comment]
@dwalend 
Reviewing the PR.

**Comment by jabrena** [comment]
Hi @dwalend,

Can you review the Unit tests about EV3Led, many tests failing now.
https://github.com/ev3dev-lang-java/ev3dev-lang-java/runs/1662912674?check_suite_focus=true

```
git checkout sysfs_perf2
./gradlew clean test --tests EV3LedTest
```

But if you compare with the previous implementation, they are running fine:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/actions/runs/465932771

```
git checkout 735
./gradlew clean test --tests EV3LedTest
```

@JakubVanek Thanks for the comments in the PR, but it is not necessary. Remember for the next time.

**Comment by jabrena** [comment]
@dwalend,

I was thinking about the testing model and how to improve it.

Currently in `Battery`
https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/sysfs_perf2/src/main/java/ev3dev/sensors/Battery.java

```
        voltageRereader = new DataChannelRereader(batteryPathLocal + "/" + voltage);
        currentRereader = new DataChannelRereader(batteryPath + "/" + batteryEv3 + "/" + current);
```

In the constructor, we initialise some Readers, but if that readers come from a Static Factory class, it could be possible to Mock it with Mockito.

Take a look this article:
https://asolntsev.github.io/en/2020/07/11/mockito-static-methods/

Juan Antonio

**Comment by jabrena** [comment]
@dwalend 

It was missed in the PR review that I didn´t find a Unit Test for the class `DataChannelRewriter` take a look the last commit in the branch, I removed the Lombok annotation `SneakyThrows`

**Comment by dwalend** [comment]
> Can you review the Unit tests about EV3Led, many tests failing now.
> https://github.com/ev3dev-lang-java/ev3dev-lang-java/runs/1662912674?check_suite_focus=true

@jabrena - that's why I've asked for some help with the test infrastructure. Those files need to be in place earlier in the tests - 
 when the objects to be tested get constructed. How should I change the code to get them there earlier? I don't follow how the current code works and need your help.

**Comment by jabrena** [comment]
No worries,

Tomorrow, I will review in detail the tests and code, in order to fix it and I will do later an explanation here.
If after the explanation, you continue with doubts, we could do a Skype or something.

Tomorrow, I will continue.

Juan Antonio

**Comment by jabrena** [comment]
Hi @dwalend

I am working in the package `fake_ev3dev` to improve the logs and `Java docs`.

**Comment by jabrena** [comment]
I found the issue:

In the support for EV3Led, the keys to get access to ev3 leds are absolutes:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/master/src/main/resources/stretch.properties

```
#LED
ev3.led.left.red=/sys/class/leds/led0:red:brick-status/brightness
ev3.led.left.green=/sys/class/leds/led0:green:brick-status/brightness
ev3.led.right.red=/sys/class/leds/led1:red:brick-status/brightness
ev3.led.right.green=/sys/class/leds/led1:green:brick-status/brightness
```

And later, if you execute the tests, EV3Led call a path that doesn´t exist in local.

``` java
if (direction == Direction.LEFT) {
    redWriter = new DataChannelRewriter(ev3DevProperties.getProperty("ev3.led.left.red"));
    greenWriter = new DataChannelRewriter(ev3DevProperties.getProperty("ev3.led.left.green"));
} else {
    redWriter = new DataChannelRewriter(ev3DevProperties.getProperty("ev3.led.right.red"));
    greenWriter = new DataChannelRewriter(ev3DevProperties.getProperty("ev3.led.right.green"));
}
```    

This is the reason that in other parts of the project, exist this utility:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/master/src/main/java/ev3dev/hardware/EV3DevFileSystem.java

``` java
EV3DevFileSystem.getRootPath();
```

In the Brick, the method will return `/sys/class` but in local if the tests enable this one:

``` java
System.setProperty(EV3DevFileSystem.EV3DEV_TESTING_KEY, FakeBattery.EV3DEV_FAKE_SYSTEM_PATH);
```

The method will return something like:

```
/var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
```

I will create a Fake EV3Led class for you tomorrow and I will review the tests.

Juan Antonio

**Comment by jabrena** [comment]
Working in the tests.

**Comment by jabrena** [comment]
Hi @dwalend,

The tests and Pipeline was fixed
https://github.com/ev3dev-lang-java/ev3dev-lang-java/actions/runs/474291430

You should have any issue when you run in local the code:

```
./gradlew clean test checkstyleMain
```

I will explain one tests from the Test class: 
https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/sysfs_perf2/src/test/java/ev3dev/actuators/ev3/EV3LedTest.java

with the test:

```java
    @Before
    public void resetTest() throws IOException {

        System.setProperty(EV3DevFileSystem.EV3DEV_TESTING_KEY, FakeBattery.EV3DEV_FAKE_SYSTEM_PATH);

        FakeBattery.resetEV3DevInfrastructure();
        new FakeBattery(EV3DevPlatform.EV3BRICK);
    }

    @Test
    public void given_actuator_when_getDirection_then_Ok() throws Exception {

        //Given
        FakeLed fakeLed = new FakeLed(EV3DevPlatform.EV3BRICK);

        //When
        @SuppressWarnings("deprecation") EV3Led led = new EV3Led(EV3Led.RIGHT);
        EV3Led led2 = new EV3Led(EV3Led.Direction.RIGHT);

        //Then
        EV3Led.Direction expectdedDirection = EV3Led.Direction.RIGHT;
        then(led.getDirection()).isEqualTo(expectdedDirection);
        then(led2.getDirection()).isEqualTo(expectdedDirection);
    }
```

In the beginning of the tests, we set a system property in order to indicate the library that we are in Testing Mode.

```
    @Before
    public void resetTest() throws IOException {

        System.setProperty(EV3DevFileSystem.EV3DEV_TESTING_KEY, FakeBattery.EV3DEV_FAKE_SYSTEM_PATH);

        FakeBattery.resetEV3DevInfrastructure();
        new FakeBattery(EV3DevPlatform.EV3BRICK);
    }
```

```
2021-01-09 17:41:12 [Test worker] INFO  fake_ev3dev.BaseElement - Switching to EV3Dev testing infrastructure
2021-01-09 17:41:12 [Test worker] TRACE fake_ev3dev.BaseElement - Removing previous fileSystem infrastructure
2021-01-09 17:41:12 [Test worker] DEBUG fake_ev3dev.BaseElement - Created Fake file system to implement devices for testing
2021-01-09 17:41:12 [Test worker] TRACE fake_ev3dev.BaseElement - Path created: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
2021-01-09 17:41:12 [Test worker] INFO  fake_ev3dev.BaseElement - 
2021-01-09 17:41:12 [Test worker] INFO  f.ev3dev.sensors.FakeBattery - Adding a Battery device in the FileSystem for: EV3BRICK
2021-01-09 17:41:12 [Test worker] TRACE fake_ev3dev.BaseElement - Creating path: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system/power_supply/lego-ev3-battery
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.utils.Shell - Command: tree /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
2021-01-09 17:41:13 [Test worker] INFO  f.ev3dev.sensors.FakeBattery - /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
└── power_supply
    └── lego-ev3-battery
        ├── current_now
        └── voltage_now

2 directories, 2 files
```

A fake object as `Martin Fowler` says is an implementation of the real behaviour from `EV3Dev`.

> Fake objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an InMemoryTestDatabase is a good example).

https://martinfowler.com/bliki/TestDouble.html

Why in the beginning of the test, we use FakeBattery? Because we use the path from battery to identify the Brick (EV3, BrickPi or PiStorm) and this is the reason that I recreate a Battery with the minimum files to run the whole system:

```
/var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
└── power_supply
    └── lego-ev3-battery
        ├── current_now
        └── voltage_now
```

When the test begin, we initialize a Fake Object to represent a Led:

``` java
        //Given
        FakeLed fakeLed = new FakeLed(EV3DevPlatform.EV3BRICK);
```

```
2021-01-09 17:41:13 [Test worker] TRACE fake_ev3dev.BaseElement - Creating path: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system/leds/led1:green:brick-status
2021-01-09 17:41:13 [Test worker] TRACE fake_ev3dev.BaseElement - Creating path: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system/leds/led1:red:brick-status
2021-01-09 17:41:13 [Test worker] TRACE fake_ev3dev.BaseElement - Creating path: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system/leds/led0:green:brick-status
2021-01-09 17:41:13 [Test worker] TRACE fake_ev3dev.BaseElement - Creating path: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system/leds/led0:red:brick-status
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.utils.Shell - Command: tree /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
2021-01-09 17:41:13 [Test worker] INFO  f.ev3dev.actuators.ev3.FakeLed - /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
├── leds
│   ├── led0:green:brick-status
│   │   └── brightness
│   ├── led0:red:brick-status
│   │   └── brightness
│   ├── led1:green:brick-status
│   │   └── brightness
│   └── led1:red:brick-status
│       └── brightness
└── power_supply
    └── lego-ev3-battery
        ├── current_now
        └── voltage_now

7 directories, 6 files
```

Once the testing environment has a minimalistic representation to operate, you use your normal object in a natural way:

```java
        LED led2 = new EV3Led(EV3Led.Direction.LEFT);
```

```
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevDistros - Providing an EV3DevDistros instance
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevDistros - Detecting EV3Dev Distro
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.utils.Shell - Command: cat /etc/os-release
2021-01-09 17:41:13 [Test worker] WARN  ev3dev.hardware.EV3DevDistros - Failed to detect distro, falling back to Stretch.
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevDistros - Debian Stretch detected
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevPlatforms - Providing a EV3DevPlatforms instance
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevDistros - Providing an EV3DevDistros instance
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevPlatforms - Detecting platform with the battery approach
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevFileSystem - ROOT_PATH modified: /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system
2021-01-09 17:41:13 [Test worker] TRACE ev3dev.utils.Sysfs - ls /var/folders/ky/khzwpw5j05l6hr8l08tnyt9m0000gn/T/ev3dev_fake_system/power_supply/lego-ev3-battery
2021-01-09 17:41:13 [Test worker] TRACE ev3dev.hardware.EV3DevPlatforms - Detected platform: EV3BRICK
2021-01-09 17:41:13 [Test worker] DEBUG ev3dev.hardware.EV3DevDistros - Providing an EV3DevDistros instance
```

Maybe, I give some light about the way of testing. Tell me if you have any doubts.

With that idea, You could use a FakeBattery to expose few files in the testing environment:

```
String ledPath = EV3DevFileSystem.getRootPath();
```

and Add the missing tests for the new classes:

- DataChannelRereader
- DataChannelRewriter        

Juan Antonio

**Comment by dwalend** [comment]
Thanks! I knew there had to be something like that in place somewhere, but didn't see where the file system slipped in. I would have never thought to look at NXTTemperatureSensorTest.java for an example.

Sorry for going a few days without a response. Kid stuff over the weekend, then I opened a new project at work.

Any reason to ever clean up things in an @After ? 

I guess my main criticisms are 1) that FakeBattery isn't a mock of the Battery class, but a mock of the battery as it appears in the file system. and 2) That's a lot of complexity of inheritance layers to mock what is presented as a file system. I think both of those are just style choices, provided we don't fall prey to the ossification it can cause.

Should I go for the EV3LargeRegulatedMotor or the EV3GyroSensor next?

EV3LargeRegulatedMotor's complexity is all packed into BaseRegulatedMotor . That'll sweep up EV3MediumRegulatedMotor and NXTRegulatedMotor in the same pass. My wayfinder branch has 4 Rereaders and 5 Rewriters, which seems like a lot. Also, I don't have an NXTRegulatedMotor to test.

EV3GyroSensor inherits from BaseSensor which isn't so bad. I'd like to eventually add a Rereader and a Rewriter for the MODE and SENSOR_MODE, but that doesn't happen in my control loop. The complexity comes in via three GenericMode s . GenericMode needs an array of Rereaders, one per sample. That changes how 6 different sensors would work in one pass. One way around that would be to make specific modes for GyroSensor. That seems a natural path to follow, but would be an expansion of scope.  

What do you think makes sense for a next step?

**Comment by jabrena** [comment]
Good morning @dwalend 

> Sorry for going a few days without a response. Kid stuff over the weekend, then I opened a new project at work.

No worries. At the end, we are investing our spare time on it. We participate because we like it. :)

**Testing:**

> Thanks! I knew there had to be something like that in place somewhere, but didn't see where the file system slipped in. I would have never thought to look at NXTTemperatureSensorTest.java for an example.

It is a new Test for a new Sensor added last week, so I implemented with the latest approach.

> Any reason to ever clean up things in an @after ?

I will review that point this week. 
I will review it to remove the temporal folder.

> I guess my main criticisms are 1) that FakeBattery isn't a mock of the Battery class, but a mock of the battery as it appears in the file system. and 2) That's a lot of complexity of inheritance layers to mock what is presented as a file system. I think both of those are just style choices, provided we don't fall prey to the ossification it can cause.

At the end the whole model tries to replace the real behaviour of EV3Dev in order to be consumed by Java.

**What do you think makes sense for a next step?**

- EV3GyroSensor

But before adding that support, remember that it is missing the following tests:

- DataChannelRereader
- DataChannelRewriter

**Note:** Review the following docs. I wrote them thinking in a classroom in mind. Review if you consider that you could reuse in your course. If you find something missing, we could improve them in the future.
https://ev3dev-lang-java.github.io/docs/support/index.html

From my side, I will be working on:
https://github.com/ev3dev-lang-java/ev3dev-lang-java/issues/771

Have a good week.

**Comment by dwalend** [comment]
I added a test of DataChannelRewriter in https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/794 .

What else would you like to see tested in DataChannelRereader ?

**Comment by jabrena** [comment]
Hi @dwalend,

It would be nice to have the same kind of test developed for DataChannelRewriter in DataChannelRereader please

**Comment by dwalend** [comment]
@jabrena - it's been merged for some time: https://github.com/ev3dev-lang-java/ev3dev-lang-java/blob/sysfs_perf2/src/test/java/ev3dev/utils/DataChannelRereaderTest.java (You've improved it. ... It's not particularly memorable, a lot smaller than the rest of your cleanup.)

**Comment by jabrena** [comment]
Forget my last message... Sometimes I write in automatic mode... I will merge it today :)

**Comment by dwalend** [comment]
@JakubVanek @jabrena  - what do you think of the approach in https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/795 ? I avoided changing GenericMode by putting a specific mode in EV3GyroSensor . The next step down that path would be two more specific modes inside of EV3GyroSensor, and probably a refactoring to get a generic "1 value" specific mode.

**Comment by dwalend** [comment]
@JakubVanek @jabrena - Juan asked me to show what the fix for EV3GyroSensor would be inside of GenericMode. It'd be https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/796 . What do you think of that one?

I'm hesitant to make changes inside GenericMode. I think my main complaint is that it touches a lot more of the system. The interaction with NXTTemperatureSensor seems particularly problematic.

The more I work inside the SampleProvider hierarchy the less comfortable I am with teaching it to kids. I'd much prefer to flatten it out to give specific APIs - for example - EV3GyroSensor a readAngle() method - and throw an exception if the sensor is in the wrong mode. That's a bigger change, for a different day, but much easier to explain and use. 

I'm much more comfortable with the path forward in https://github.com/ev3dev-lang-java/ev3dev-lang-java/pull/795 , which limits the change to EV3GyroSensor . (Just one mode of the three in this pull request.)

**Comment by jabrena** [comment]
Hi @dwalend,

> The more I work inside the SampleProvider hierarchy the less comfortable I am with teaching it to kids. 

This is issue is not add new methods, the issue is to increase the performance with the new classes. I understand it, but at the moment, it is not possible. I will fix today/tomorrow the failed tests. 

Upgrading GenericMode with the new approach, we upgrade many parts of the system, so it is not a problem.

**Comment by dwalend** [comment]
If we're to follow that same path for the motors then I can put together changes in BaseRegulatedMotor that will take care of both of the EV3 motors. 

That's 4 DataChannelRereader s and 5 DataChannelRewriter s in one go. That OK?

**Comment by jabrena** [comment]
Hi @dwalend,

Ok, go ahead with that changes

**Comment by jabrena** [comment]
Hi @dwalend,

Returning to the issue.
I was busy with the university.

and you?

Juan Antonio
